pkg/api/types.go

```diff
@@ -688,36 +691,66 @@ type ServiceStatus struct{}
 
 // ServiceSpec describes the attributes that a user creates on a service
 type ServiceSpec struct {
-	// Port is the TCP or UDP port that will be made available to each pod for connecting to the pods
-	// proxied by this service.
-	Port int `json:"port"`
+	// Required: The list of ports that are exposed.
+	// FIXME: this would be nicer as a map of name -> struct
+	Ports []ServicePort
 
-	// Optional: Supports "TCP" and "UDP".  Defaults to "TCP".
-	Protocol Protocol `json:"protocol,omitempty"`
+	// Either 'Selector' or 'Endpoints' must be set:
 
-	// This service will route traffic to pods having labels matching this selector. If empty or not present,
-	// the service is assumed to have endpoints set by an external process and Kubernetes will not modify
-	// those endpoints.
+	// This service will route traffic to pods having labels matching this
+	// selector.
 	Selector map[string]string `json:"selector"`
 
+	// Use the specified endpoints.
+	// FIXME: Some weird interaction.  All the fields of Ports[] are basically
+	//   ignored in favor of Endpoints.  Maybe make this just a list of IPs,
+	//   and disallow named DestinationPorts if using this mode?
+	Endpoints []Endpoint
+
 	// PortalIP is usually assigned by the master.  If specified by the user
 	// we will try to respect it or else fail the request.  This field can
 	// not be changed by updates.
 	PortalIP string `json:"portalIP,omitempty"`
 
-	// CreateExternalLoadBalancer indicates whether a load balancer should be created for this service.
+	//FIXME: Should these next 3 be per-port (different external IPs?) or for all ports?
+	// CreateExternalLoadBalancer indicates whether a load balancer should
+	// be created for this service.
 	CreateExternalLoadBalancer bool `json:"createExternalLoadBalancer,omitempty"`
+
 	// PublicIPs are used by external load balancers.
 	PublicIPs []string `json:"publicIPs,omitempty"`
```

Comment:
> On Feb 14, 2015, at 5:03 PM, Tim Hockin <notifications@github.com> wrote:
> 
> In pkg/api/types.go:
> 
> >  	CreateExternalLoadBalancer bool `json:"createExternalLoadBalancer,omitempty"`
> > +
> >  	// PublicIPs are used by external load balancers.
> >  	PublicIPs []string `json:"publicIPs,omitempty"`
> We could rename if that brings clarity. I was trying to resist really digging into this area, as much as I don't like it, for the sake of making forward progress.
> 
> Forcing two services to get internal and external sounds OK at first blush, but I'm not 100% sure. I think we can easily follow up with that, but I'd like to get this aspect (ports) locked down first.
> 
> @smarterclayton users can set PublicIP for cases where their cloud provider does not support createExternalLoadBalancer. I want to change ALL of this, and Brian and Joe and I have been bouncing ideas around about it (as you saw on one of the issues, but more IRL, too), but I don't think plural ports need to block on it.
> 
Was not suggesting it should.  I don't think we're talking about the same users though -a cluster admin (small cluster, 2 nodes) should be able to make all services route.  However, the larger a cluster gets the likelihood an end user wants or should be allowed to define external routing drops precipitously (or at least, I haven't heard a good case for it).  Setting a public ip for a service is not a standard app creator use case.

> —
> Reply to this email directly or view it on GitHub.
>